system-addon/test/unit/lib/Screenshots.test.js

```diff
@@ -49,6 +49,34 @@ describe("Screenshots", () => {
     });
   });
 
+  describe("#maybeCacheScreenshot", () => {
+    let link;
+    beforeEach(() => {
+      link = {__sharedCache: {updateLink: sinon.stub()}};
+      sandbox.stub(Screenshots, "getScreenshotForURL");
+    });
+    it("should call getScreenshotForURL", () => {
+      Screenshots.maybeCacheScreenshot(link, "mozilla.com", "image", sinon.stub());
+
+      assert.calledOnce(Screenshots.getScreenshotForURL);
+      assert.calledWithExactly(Screenshots.getScreenshotForURL, "mozilla.com");
+    });
+    it("should not call getScreenshotForURL if a fetch is in progress", () => {
+      link.__sharedCache.fetchingScreenshot = sinon.stub();
+
+      Screenshots.maybeCacheScreenshot(link, "mozilla.com", "image", sinon.stub());
+
+      assert.notCalled(Screenshots.getScreenshotForURL);
+    });
+    it("should not call getScreenshotsForURL if property !== undefined", () => {
+      link.image = null;
```

Comment:
Similarly, you can test the actual bug instead of testing the implementation detail by having `getScreenshotForURL` resolve `null` and calling `maybeCacheScreenshot` a second time.

Reply:
Wouldn't that (testing by calling maybeCacheScreenshot 2 times) require a proper implementation of `updateLink`that actually sets the `image` property?

Reply:
@Mardak the comment got collapsed but this still needs to be addressed.